### PR TITLE
feat: add users/groups data sources

### PR DESCRIPTION
**Pull request by mxyng**
Add data sources for users and groups. Options include querying users by name and group membership and groups by name and members. Output include including group members for groups and groups a user belongs to.

### CODE REVIEW

File: docs/data-sources/groups.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+Optional:
+
+- `name` (String) Then name of the group to find.
+- `user_id` (String) The ID of the user who belongs to this group. Cannot be used with `user_name`.
+- `user_name` (String) The name of the user who belongs to this group. Cannot be used with `user_id`.
+
+
+<a id="nestedatt--groups"></a>
+### Nested Schema for `groups`
```
Comment (mxyng) [inline docs/data-sources/groups.md:61]: This is a bug with the doc generator. It omits read-only nested field descriptions

File: internal/provider/data_source_users.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/infrahq/infra/api"
+)
+
+func dataSourceUsers() *schema.Resource {
+	return &schema.Resource{
+		Description: "Get a list of registered Infra users.",
```
Comment (jmorganca) [inline internal/provider/data_source_users.go:17]:
```suggestion
		Description: "Get a list of Infra users.",
```

Comment (jmorganca) [inline internal/provider/data_source_users.go]:
```suggestion
		Description: "Get a list of Infra users.",
```

File: internal/provider/data_source_users.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+				Computed: true,
+			},
+			"filter": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": &schema.Schema{
+							Description:      "The name of the user to find.",
```
Comment (jmorganca) [inline internal/provider/data_source_users.go]:
```suggestion
							Description:      "The name of the user.",
```

Small nit but should these end in `.`?

File: internal/provider/data_source_groups.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/infrahq/infra/api"
+)
+
+func dataSourceGroups() *schema.Resource {
+	return &schema.Resource{
+		Description: "Get a list of registered Infra groups.",
```
Comment (jmorganca) [inline internal/provider/data_source_groups.go]:
```suggestion
		Description: "Get a list of Infra groups.",
```

File: internal/provider/data_source_groups.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+				Computed: true,
+			},
+			"filter": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": &schema.Schema{
+							Description: "Then name of the group to find.",
```
Comment (jmorganca) [inline internal/provider/data_source_groups.go]:
```suggestion
							Description: "Then name of the group.",
```

**Review by jmorganca — APPROVED** [review]
Minor syntax suggestions

File: internal/provider/data_source_groups_test.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	name := randomName()
+
+	dataSourceName := fmt.Sprintf("data.infra_groups.%s", t.Name())
+
+	resource.UnitTest(t, resource.TestCase{
+		PreCheck:          testAccPreCheck(t),
+		ProviderFactories: testAccProviders(t),
+		Steps: []resource.TestStep{
+			{
+				Config: composeTestConfigFunc(
```
Comment (mxyng) [inline internal/provider/data_source_groups_test.go:21]: We want to preload the test environment with some resources so the data sources can find them in later steps. It's necessary to keep the resources around in each step so Terraform doesn't destroy them

File: internal/provider/data_source_users.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+	for i := range d.Get("filter").([]interface{}) {
+		request.Name = d.Get(fmt.Sprintf("filter.%d.name", i)).(string)
+
+		if d.Get(fmt.Sprintf("filter.%d.group_id", i)).(string) != "" || d.Get(fmt.Sprintf("filter.%d.group_name", i)).(string) != "" {
+			group, err := groupFromIDOrName(ctx, client, d, fmt.Sprintf("filter.%d.group_id", i), fmt.Sprintf("filter.%d.group_name", i))
+			if err != nil {
+				return diag.FromErr(err)
+			}
+
+			request.Group = group.ID
```
Comment (mxyng) [inline internal/provider/data_source_users.go:106]: This could set `group_id` or `group_name`, whichever is missing

**Pull request merged**